torchdata/dataloader2/dataloader2.py

```diff
@@ -88,25 +121,16 @@ def __iter__(self) -> Iterator[T_co]:
                 self.reading_service.initialize_iteration()
 
             self._datapipe_iter = iter(self.datapipe)
-
             self._reset_iter = False
 
-        return self
-
-    def __next__(self) -> T_co:
-        if self._reset_iter:
-            raise StopIteration
-        try:
-            return next(self._datapipe_iter)  # type: ignore[arg-type]
-        except PauseIteration:
-            raise StopIteration
-        except StopIteration:
-            if self.reading_service is not None:
-                self.reading_service.finalize_iteration()
-            self._reset_iter = True
-            raise
+        self.valid_iterator_id = 0 if self.valid_iterator_id is None else self.valid_iterator_id + 1
+        return _DataLoader2Iterator(self, self.valid_iterator_id)
 
     def __getattr__(self, name: str) -> Any:
+        """
+        Delegate methods (e.g. `limit`, `pause`, `resume`, etc) to the iterator
+        created by the ReadingService and DataPipe.
+        """
```

Comment:
If we directly delegate to `datapipe_iter` from `dataloader_iter`, we don't need this function.

Reply:
I am fine with this change, but I would flag that methods such as `state_dict` and `shutdown` will not be delegated to `DataLoader2`.

Reply:
Also, do we want to allow users to do `dl2.limit()`, ``dl2.resume()`, and etc? Or we want them to always invoke those from the iterator?

If we want the former, we will need to keep this method.

Reply:
cc: @Miiira

Reply:
I think invoking from one single place (i.e. iterator) rather than multiple places is better and less error-prone.

The internal test is calling `.resume` on the iterator, so I think it is fine to remove the API from the `DataLoader2` class.

Reply:
Yes we only need `limit`, `resume` on iterator. I'm good with removing this from `DataLoader2`. Maybe we also want to look at Lightning DataModule `train_dataloader` return type with this change